### PR TITLE
SSLR-380 Add an example for the new Strongly Typed Construction API

### PR DESCRIPTION
**Pull request by vilchik-elena**

### CODE REVIEW

File: sslr-examples/src/main/java/org/sonar/sslr/examples/grammars/typed/JsonGrammar.java [hunk 1]
```diff
@@ -0,0 +1,95 @@
+package org.sonar.sslr.examples.grammars.typed;
```
Comment (dbolkensteyn) [inline sslr-examples/src/main/java/org/sonar/sslr/examples/grammars/typed/JsonGrammar.java]: the build is failing on my machine due to missing license headers

Comment (vilchik-elena) [inline sslr-examples/src/main/java/org/sonar/sslr/examples/grammars/typed/JsonGrammar.java]: @dbolkensteyn Pff, sorry. Fixed in a3530a9ee0a327ecfdb2479ddc320c6b20f4df5b

File: sslr-examples/src/main/java/org/sonar/sslr/examples/grammars/typed/JsonPunctuator.java [hunk 2]
```diff
@@ -0,0 +1,28 @@
+package org.sonar.sslr.examples.grammars.typed;
+
+import org.sonar.sslr.grammar.GrammarRuleKey;
+
+enum JsonPunctuator implements GrammarRuleKey {
```
Comment (dbolkensteyn) [inline sslr-examples/src/main/java/org/sonar/sslr/examples/grammars/typed/JsonPunctuator.java]: this class doesn't bring value on its own.

I'd simply drop it and move the `LCURLYBRACE`, `COMMA`, etc. declarations to `JsonLexer`

File: sslr-examples/src/main/java/org/sonar/sslr/examples/grammars/typed/JsonLexer.java [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+    b.rule(TRUE).is("true", WHITESPACE);
+    b.rule(FALSE).is("false", WHITESPACE);
+    b.rule(NULL).is("null", WHITESPACE);
+
+    b.rule(WHITESPACE).is(b.regexp("[ \n\r\t\f]*+"));
+    b.rule(NUMBER).is(b.regexp("-?+(0|[1-9][0-9]*+)(\\.[0-9]++)?+([eE][+-]?+[0-9]++)?+"), WHITESPACE);
+    b.rule(STRING).is(b.regexp("\"([^\"\\\\]|\\\\([\"\\\\/bfnrt]|u[0-9a-fA-F]{4}))*+\""), WHITESPACE);
+
+    b.rule(EOF).is(b.token(GenericTokenType.EOF, b.endOfInput())).skip();
```
Comment (dbolkensteyn) [inline sslr-examples/src/main/java/org/sonar/sslr/examples/grammars/typed/JsonLexer.java]: you should be able to remove the `skip()` call here

calling `skip()` and `skipIfOneChild()` when using the new strongly typed ASTs should [ideally] not be required.
indeed, these methods were useful mainly to control the shape of the `AstNode` tree itself generated by the SSLR parser

Comment (vilchik-elena) [inline sslr-examples/src/main/java/org/sonar/sslr/examples/grammars/typed/JsonLexer.java]: @dbolkensteyn done 0794da1e3757b07c1d75cd84c266bf0047b3bc6d

File: sslr-examples/src/main/java/org/sonar/sslr/examples/grammars/typed/JsonPunctuator.java (hunk 2 above)
Comment (vilchik-elena) [inline sslr-examples/src/main/java/org/sonar/sslr/examples/grammars/typed/JsonPunctuator.java]: @dbolkensteyn fixed b209900902c04aea5f89499a0740735fecf4cde2

File: sslr-examples/src/main/java/org/sonar/sslr/examples/grammars/typed/JsonNodeBuilder.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+import com.sonar.sslr.api.typed.Input;
+import com.sonar.sslr.api.typed.NodeBuilder;
+import java.util.List;
+import org.sonar.sslr.examples.grammars.typed.impl.InternalSyntaxToken;
+import org.sonar.sslr.grammar.GrammarRuleKey;
+
+public class JsonNodeBuilder implements NodeBuilder {
+
+  @Override
+  public Object createNonTerminal(GrammarRuleKey ruleKey, Rule rule, List<Object> children, int startIndex, int endIndex) {
```
Comment (dbolkensteyn) [inline sslr-examples/src/main/java/org/sonar/sslr/examples/grammars/typed/JsonNodeBuilder.java]: this part is a bit tricky and of black magic - perhaps it deserves a comment

File: sslr-examples/src/main/java/org/sonar/sslr/examples/grammars/typed/JsonNodeBuilder.java [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+    for (Object child : children) {
+      if (child instanceof InternalSyntaxToken) {
+        return child;
+      }
+    }
+    return null;
```
Comment (dbolkensteyn) [inline sslr-examples/src/main/java/org/sonar/sslr/examples/grammars/typed/JsonNodeBuilder.java]: i think this is unreachable - or at least is not expected to be reached

File: sslr-examples/src/main/java/org/sonar/sslr/examples/grammars/typed/JsonNodeBuilder.java [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+  }
```
Comment (dbolkensteyn) [inline sslr-examples/src/main/java/org/sonar/sslr/examples/grammars/typed/JsonNodeBuilder.java]: so this methods seems to be called for every rule defined in `JsonLexer` (i.e. `TRUE`, `NUMBER`, etc.)

these are supposed to be tokens in fact (as `createTerminal`), but are non-terminals due to whitespaces.
the whitespace token is dropped, and only the first one is returned

**Comment by dbolkensteyn** [comment]
LGTM - I'd just slightly change `JsonNodeBuilder` because that's the hardest part to understand

Otherwise it's good to go, feel free to merge & to respond on the google group thread

**Pull request merged**